### PR TITLE
Add `binding` pkg

### PR DESCRIPTION
**Pull request by mysqto**
Purpose:
1. Make `gin` related bindings generic
2. Update `go.mod`

### CODE REVIEW

File: http/binding/binding_test.go [hunk 1]
```diff
@@ -0,0 +1,472 @@
...
+	})
+	s.router.PATCH(testJSONEndpoint, testJSONHandler)
+	r := httptest.NewRecorder()
+	s.router.ServeHTTP(r, req)
+
+	s.Equal(http.StatusOK, r.Code)
+
+	body, err := json.Marshal(&testReq)
+	s.NoError(err)
+	s.Equal(r.Body.String(), string(body))
```
Comment (viettd) [inline http/binding/binding_test.go:160]: I think `body` is not what in the context

File: http/binding/binding_test.go [hunk 2]
```diff
@@ -0,0 +1,472 @@
...
+	req, err := http.NewRequest(http.MethodPatch, testJSONEndpoint, strings.NewReader(requestBody))
+	s.NoError(err)
+
+	s.router.PATCH(testJSONEndpoint, testJSONHandler)
+	r := httptest.NewRecorder()
+	s.router.ServeHTTP(r, req)
+
+	s.Equal(http.StatusOK, r.Code)
+
+	s.Equal(r.Body.String(), "{\"number\":[1,2,3],\"string\":[\"A\",\"B\",\"C\"]}")
```
Comment (viettd) [inline http/binding/binding_test.go:195]: For type mismatch, I think it should bind from the body right?
But I see it returns the value from the context here.

File: http/binding/binding_test.go (hunk 1 above)
Comment (mysqto) [reply http/binding/binding_test.go:160]: This case is try to test we already have the request in context, the value are set here
https://github.com/wego/pkg/blob/a07d96900f5ed59ba1fe46be2810be24a5fe5435/http/binding/binding_test.go#L148-L151 @viettd

File: http/binding/binding_test.go (hunk 2 above)
Comment (mysqto) [reply http/binding/binding_test.go:195]: The value in context is `testReq`(1234 and ABCD),  the actual value is `requestBody`(123 and ABC), here the result should be equal to `requestBody`

File: http/binding/binding_test.go [hunk 3]
```diff
@@ -0,0 +1,472 @@
...
+}
+
+func (s *BindingSuite) Test_BindQuery_FromBodyBindError() {
```
Comment (viettd) [inline http/binding/binding_test.go:198]: Shall we rename those test cases to `Test_BindQuery_FromURL` or something similar?
`FromBody` is not correct for these cases.

File: http/binding/binding_test.go [hunk 4]
```diff
@@ -0,0 +1,472 @@
...
+	})
+	s.router.POST(testQueryEndpoint, testQueryHandler)
+	r := httptest.NewRecorder()
+	s.router.ServeHTTP(r, req)
+
+	s.Equal(http.StatusOK, r.Code)
+
+	body, err := json.Marshal(&testReq)
+	s.NoError(err)
+	s.Equal(r.Body.String(), string(body))
```
Comment (viettd) [inline http/binding/binding_test.go:260]: Same as https://github.com/wego/pkg/pull/29/files#r739973486

File: http/binding/binding_test.go (hunk 1 above)
Comment (viettd) [reply http/binding/binding_test.go:160]: Please ignore, my mistake 😂

File: http/binding/binding_test.go (hunk 2 above)
Comment (viettd) [reply http/binding/binding_test.go:195]: Please ignore 😂

File: http/binding/binding_test.go (hunk 1 above)
Comment (mysqto) [reply http/binding/binding_test.go:160]: I changed the `testReq` to `inCtx`, make it easier to understand

**Review by viettd — APPROVED** [review]

**Pull request merged**